src/jailer/src/cgroup.rs

```diff
@@ -1,25 +1,224 @@
+}
+
+impl FromStr for CgroupVersion {
+    type Err = Error;
+
+    fn from_str(s: &str) -> std::result::Result<CgroupVersion, Self::Err> {
+        match s {
+            "1" => Ok(CgroupVersion::V1),
+            "2" => Ok(CgroupVersion::V2),
+            _ => Err(Error::CgroupInvalidVersion(s.to_string())),
+        }
+    }
+}
+
+// Holds information on a cgroup mount point discovered on the system
+struct CgroupMountPoint {
+    dir: String,
+    options: String,
+}
+
+// Allows creation of cgroups on the system for both versions
+pub struct CgroupBuilder {
+    version: CgroupVersion,
+    hierarchies: HashMap<String, PathBuf>,
+    mount_points: Vec<CgroupMountPoint>,
+}
+
+impl CgroupBuilder {
+    // Creates the builder object
+    // It will discover cgroup mount points and hierarchies configured
+    // on the system and cache the info required to create cgroups later
+    // within this hierarchies
+    pub fn new(ver: CgroupVersion) -> Result<Self> {
+        let mut b = CgroupBuilder {
+            version: ver,
+            hierarchies: HashMap::new(),
+            mount_points: Vec::new(),
+        };
+
+        // search PROC_MOUNTS for cgroup mount points
+        let f =
+            File::open(PROC_MOUNTS).map_err(|e| Error::FileOpen(PathBuf::from(PROC_MOUNTS), e))?;
+
+        // Regex courtesy of Filippo.
+        // This will match on both v1 and v2 mount points.
+        // For v1, the "ver" capture group will be empty (len = 0)
+        let re = Regex::new(
+            r"^([a-z2]*)[[:space:]](?P<dir>.*)[[:space:]]cgroup(?P<ver>2?)[[:space:]](?P<options>.*)[[:space:]]0[[:space:]]0$",
+        ).map_err(Error::RegEx)?;
+
+        let mut found = false;
+        for l in BufReader::new(f).lines() {
+            let l = l.map_err(|e| Error::ReadLine(PathBuf::from(PROC_MOUNTS), e))?;
+            if let Some(capture) = re.captures(&l) {
+                if ver == CgroupVersion::V2 && capture["ver"].len() == 1 {
+                    // Found the cgroupv2 unified mountpoint; with cgroupsv2 there is only one
+                    // hierarchy so we insert it in the hashmap to use it later when creating
+                    // cgroups
+                    b.hierarchies
+                        .insert("unified".to_string(), PathBuf::from(&capture["dir"]));
+                    found = true;
+                    break;
+                } else if ver == CgroupVersion::V1 && capture["ver"].is_empty() {
+                    // Found a cgroupv1 mountpoint; with cgroupsv1 we can have multiple hiearachies.
+                    // Since we don't know which one will be used, we cache the mountpoints now,
+                    // and will create the hierachies on demand when a cgroup is built.
+                    b.mount_points.push(CgroupMountPoint {
+                        dir: String::from(&capture["dir"]),
+                        options: String::from(&capture["options"]),
+                    });
+                    found = true;
```

Comment:
Should we break loop here ?

Reply:
No. Different cgroup-v1 hierarchies can be mounted at different locations. This is actually the standard approach.
We break the loop in case of Cgroup-v2 because there can be a single mount point in the system and there is no point in searching further.